keps/sig-api-machinery/20190426-crd-defaulting.md

```diff
@@ -0,0 +1,336 @@
+
+We do this in the serializer by passing a real defaulter to [`versioningserializer.NewCodec`](https://github.com/kubernetes/apimachinery/blob/master/pkg/runtime/serializer/versioning/versioning.go#L49) such that defaulting is done natively just after the binary payload has been unmarshalled into an `map[string]interface{}` and pruning of [KEP: Pruning for CustomResources](/keps/sig-api-machinery/20180731-crd-pruning.md) was done.
+
+Like for native resources, we do defaulting
+
+* during request payload deserialization
+* after mutating webhook admission
+* during read from storage.
+     
+Note: like for native resources, we do not default after webhook conversions. Hence, webhook conversions should be complete in the sense that they return defaulted objects in order for the API user to see defaulted objects. Technically we could do additional defaulting, but to match native resources, we do not.
+
+Compare the yellow boxes in the following figure:
+
+![Decoding steps which must apply defaults](20190426-crd-defaulting-pipeline.png)
+
+We rely on the validation steps in the request pipeline to verify that the default value validates value validation. We will check the types in default values using the _structural schema_ during CRD creation and update though. We will also reject defaults which contain values which will be pruned.
+
+The `default` field in the CRD types is considered alpha quality. We will add a `CustomResourceDefaulting` feature gate. Values for `default` will be rejected if the gate is not enabled and there have not been `default` values set before (ratcheting validation). 
+
+[Kubebuilder's crd-gen](https://github.com/kubernetes-sigs/controller-tools/tree/master/cmd/crd) can make use of this feature by adding another tag, e.g. `// +default=<arbitrary-json-value>`. Defaults are arbitrary JSON values, which must also validate (types are checked during CRD creation and update, value validation is checked for requests, but not for etcd reads) and are not subject to pruning (defaulting happens after pruning).
+
+### Examples
+
+1. default in the undefined case
+
+   ```yaml
+   type: object
+   properties:
+     foo:
+       type: string
+       default: "abc"
+   ```
+
+   Then
+   
+   ```json
+   {}
+   ```
+   
+   is defaulted to:
+   
+   ```json
+   {
+     "foo": "abc"
+   }
+   ```
+   
+2. no defaulting
+
+   ```yaml
+   type: object
+   properties:
+     foo:
+       type: string
+       default: "abc"
+   ```
+
+   Then
+   
+   ```json
+   {
+     "foo": "def"
+   }
+   ```
+   
+   is defaulted to:
+   
+   ```json
+   {
+     "foo": "def"
+   }
+   ```
+   
+3. default in the undefined case
+
+   ```yaml
+   type: object
+   properties:
+     foo:
+       type: array
+       items:
+         type: integer
+       default: [1]
```

Comment:
I'd be interested in seeing an example that is more complex than just a scalar. Can I have a default build an object?

Reply:
this is an array, no scalar. Object looks the same: {...}

Reply:
I meant, is there a situation where you'd get
```
default: {"field": "value", "list": [1]}
```

Reply:
Sure. Default can hold arbitrary JSON

Reply:
if the default json creates an object that doesn't have all the default set, are they going to also be defaulted?

Reply:
They're applied top-down, so I would expect so, but that's good to spec and test

Reply:
top-down, yes. Adding a sentence.

Reply:
added an example as well.